src/viam/sdk/resource/reconfigurable.hpp

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/resource/resource.hpp>
+
+namespace viam {
+namespace sdk {
+
+class Reconfigurable {
+   public:
+    virtual ~Reconfigurable();
+
+    /// @brief Reconfigures a resource.
+    /// @param deps Dependencies of the resource.
+    /// @param cfg The resource's config.
+    virtual void reconfigure(const Dependencies& deps, const ResourceConfig& cfg) = 0;
+
+    /// @brief Reconfigures a resource if it is Reconfigurable.
+    /// @param resource the Resource to reconfigure.
+    /// @param deps Dependencies of the resource.
+    /// @param cfg The resource's config.
+    static void reconfigure_if_reconfigurable(const std::shared_ptr<Resource>& resource,
```

Comment:
Not sure if we discussed this before or not, but I wonder if you could just make this a non-virtual method of `Resource`:

```
class Resource {
    ...
    void reconfigure(const Dep..., const Res...);
    ...
};

void Resource::reconfigure(...) {
    auto reconfigurable_res = std::dynamic_pointer_cast<Reconfigurable>(resource);
    if (reconfigurable_res) {
        reconfigurable_res->reconfigure(...);
    }
```

Same could be done for `stop` / `Stoppable`?

Reply:
I like this, it seems nice to have `stop` and `reconfigure` sit on the `Resource` class. Happy enough to make the change here, it seems pretty simple.

One clarifying point, I _believe_ the methods will have to be `static` because otherwise the `this` arg is a raw pointer, but `dynamic_pointer_cast` requires a `shared_ptr`. Is there some cleverness I'm missing for how to have a non-static member method allow for a `dynamic_pointer_cast` of `this`?

Reply:
Oh, sorry - I hadn't really focused on the fact that it took a `shared_ptr` arg. But I don't think it is an obstacle: you can just use plain `dynamic_cast` instead of `dynamic_pointer_cast`.